core/src/main/scala/com/softwaremill/react/kafka_api_opt3/Example.scala

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package com.softwaremill.react.kafka_api_opt3
+
+import scala.concurrent.Future
+
+import akka.Done
+import com.softwaremill.react.kafka_api_producer.Producer
+import org.apache.kafka.clients.consumer.KafkaConsumer
+import org.apache.kafka.clients.producer.KafkaProducer
+import org.apache.kafka.clients.producer.ProducerRecord
+
+trait Example {
+  implicit val ec = scala.concurrent.ExecutionContext.Implicits.global
+
+  def consumerProvider: () => KafkaConsumer[Array[Byte], String] = ???
+
+  def producerProvider: () => KafkaProducer[Array[Byte], String] = ???
+
+  class DB {
+    def save(offset: Long, data: String): Future[Done] = ???
+  }
+}
+
+// Consume messages and store a representation, including offset, in DB
+object Example3 extends Example {
+  val db = new DB
+
+  Consumer.source(consumerProvider)
```

Comment:
> I don't see any deadlock here. Can you build an example that deadlocks? The onComplete will be emitted, just after last message has been committed (at least in the scenario I described).

```scala
//pseudo code
Source
  .splitAfter()
  .map()
  .fold()
  .mergeSubstreams
  .map(/* commit here */)
```

Without emitting `onComplete` `splitAfter` will not stop and will hold some not processed and not committed elements.

Also `sliding` may hold some elements that will never be emitted at all.

> Then I propose the following external stop() semantics:

I dont like put timeout into `Source`. IMO owner of `Source` should decide and implement shutdown process. All `Source` should do is try to shutdown.

This will always fail in case of some messages committed and some no (usual for kafka, because kafka operates offsets, not messages) and in case when messages not committed at all (external offset storing).